### PR TITLE
Minor improvements for data reading and caching

### PR DESCRIPTION
**Pull request by Martin-Molinero**
#### Description

- Minor improvements for data reading and caching
    - `ZipDataCacheProvider` will now support reading and writing at the same time in a thread safe manner
    - Replace `bool` flags for thread (& memory/cache) safe operations
- LeanDataReader will now use stream Reader implementations if available, for speed
- Fix LeanDataReader bug where would not be able to determine the correct TickType from file name
#### Related Issue

N/A
#### Motivation and Context

Low resolution aggregation
#### Requires Documentation Change

N/A
#### How Has This Been Tested?

Running lean data reader and writer
#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

**Review by jaredbroad — APPROVED** [review]

**Review by jaredbroad — APPROVED** [review]

File: Common/Util/LeanData.cs [hunk 1]
```diff
@@ -1013,7 +1013,12 @@ public static bool TryParseSecurityType(string fileName, out SecurityType securi
                 var fileName = Path.GetFileNameWithoutExtension(filePath);
                 if (fileName.Contains("_"))
                 {
-                    tickType = (TickType)Enum.Parse(typeof(TickType), fileName.Split('_')[1], true);
+                    var tickTypePosition = 1;
```
Comment (C-SELLERS) [inline Common/Util/LeanData.cs:1019]: 😔, hate that this is so niche and specific. Would be good to have examples documented of each type of filename format.

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 2]
```diff
@@ -260,18 +277,30 @@ private Stream CacheAndCreateEntryStream(string filename, string entryName)
...
-                entry = zipFile.EntryCache.FirstOrDefault().Value;
+                entryCache = zipFile.EntryCache.FirstOrDefault().Value;
             }
             else
             {
-                zipFile.EntryCache.TryGetValue(entryName, out entry);
+                zipFile.EntryCache.TryGetValue(entryName, out entryCache);
+            }
+
+            if (entryCache is { Modified: true })
```
Comment (C-SELLERS) [inline Engine/DataFeeds/ZipDataCacheProvider.cs:292]: 🔥, Seems like this should address the issue I was having where we were attempting to read an entry after a `Store()`

File: Common/Util/LeanData.cs (hunk 1 above)
Comment (Martin-Molinero) [reply Common/Util/LeanData.cs:1019]: yeah, it's all centralized logic now in `LeanData`, covered by the new unit tests too 👍

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 3]
```diff
@@ -467,13 +500,20 @@ public void Dispose()
...
                 }
-
-                Disposed = true;
             }
         }
+
+        /// <summary>
+        /// ZipEntry wrapper which handles flagging a modified entry
+        /// </summary>
+        private class ZipEntryCache
```
Comment (C-SELLERS) [inline Engine/DataFeeds/ZipDataCacheProvider.cs:513]: Cool, making it entry specific is nice rather then per zip

File: ToolBox/LeanDataReader.cs [hunk 4]
```diff
@@ -114,10 +108,40 @@ public LeanDataReader(string filepath)
...
         {
+            if (!File.Exists(_zipPath))
+            {
+                Log.Error($"LeanDataReader.Parse(): File does not exist: {_zipPath}");
+                yield break;
+            }
+
             var factory = (BaseData) ObjectActivator.GetActivator(_config.Type).Invoke(new object[0]);
 
+            if (_config.Type.ImplementsStreamReader())
```
Comment (C-SELLERS) [inline ToolBox/LeanDataReader.cs:119]: Nice!!! 👍🏽 

Great to see the LeanDataReader support stream reading.

**Review by C-SELLERS — APPROVED** [review]
Looks good to me! 👍🏽

**Pull request merged**